### PR TITLE
Cluster modules documentation

### PR DESCRIPTION
**Pull request by libmonsoon-dev**
## Description
Added a controller that, when restarting a service with documentation, fills it with documentation for available modules

## Why do we need it, and what problem does it solve?
https://github.com/deckhouse/deckhouse/issues/6145

## What is the expected result?
Module documentation is available in cluster documentation

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: documentation
type: feature
summary: Module documentation is available in the cluster.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/openstack'...

**Comment by z9r5** [comment]
/e2e/abort static;Static;containerd;1.25 6864524263 6864524263-1-con-1-25 /sys/deckhouse-oss/install:pr6449 astra@192.168.199.145

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've recognized your '/e2e/abort' command and started the workflow...

**Review by name212 — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'deploy/web/test'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'deploy/web/test'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'deploy/web/test'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'deploy/web/test'...

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 1]
```diff
@@ -338,6 +344,11 @@ func (c *Controller) createOrUpdateReconcile(ctx context.Context, roMR *v1alpha1
 	// Or create a copy manually for better performance
 	mr := roMR.DeepCopy()
 
+	err := c.sendDocumentation(ctx, mr)
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:347]: I guess we have to build documentation only for deployed modules. So this function should be removed to the `case v1alpha1.PhaseDeployed:` like that:
```
	case v1alpha1.PhaseDeployed:
		err := c.sendDocumentation(ctx, mr)
		if err != nil {
			return ctrl.Result{Requeue: true}, fmt.Errorf("send documentation error: %v", err)
		}

		// add finalizer and status label
		if !controllerutil.ContainsFinalizer(mr, fsReleaseFinalizer) {
			controllerutil.AddFinalizer(mr, fsReleaseFinalizer)
		}

....
```

Comment (yalosev) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go:347]: Also we have to call this method in the override-controller, when we deploy the module somehere [here](https://github.com/deckhouse/deckhouse/blob/main/deckhouse-controller/pkg/controller/module-controllers/release/override-controller.go#L299)

File: deckhouse-controller/pkg/controller/module-controllers/release/controller.go [hunk 2]
```diff
@@ -932,3 +1018,66 @@ type moduleValidator interface {
...
+	if len(addrs) == 0 {
+		return nil
+	}
+
+	ms, err := c.moduleSourcesLister.Get(mr.GetModuleSource())
+	if err != nil {
+		return fmt.Errorf("get module source: %w", err)
+	}
+
+	md := downloader.NewModuleDownloader(c.externalModulesDir, ms, utils.GenerateRegistryOptions(ms))
```
Comment (miklezzzz) [inline deckhouse-controller/pkg/controller/module-controllers/release/controller.go:1037]: I'm wondering if we really have to download a release image one more time to get its documentation. We could get it from file system, couldn't we?

Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/release/controller.go:1037]: There shouldn't be any problems with this, we can do it in the next release

**Review by miklezzzz — APPROVED** [review]

**Review by z9r5 — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'e2e/run/aws'...

File: deckhouse-controller/pkg/controller/module-controllers/downloader/module_downloader.go [hunk 3]
```diff
@@ -137,6 +138,24 @@ func (md *ModuleDownloader) DownloadMetadataFromReleaseChannel(moduleName, relea
 	return res, nil
 }
 
+func (md *ModuleDownloader) FetchModuleVersionFromReleaseChannel(moduleName string, releaseChannel string) (moduleVersion string, err error) {
```
Comment (nabokihms) [inline deckhouse-controller/pkg/controller/module-controllers/downloader/module_downloader.go:141]: NIT: we try not to use named returned values. 
The better way is to explicitly return them like `return moduleVersion, err`

File: go_lib/module/docs-builder/client.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+	res, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, 0, err
+	}
+	defer res.Body.Close()
+
+	dataBytes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, 0, err
```
Comment (nabokihms) [inline go_lib/module/docs-builder/client.go:94]: NIT
```suggestion
		return nil, res.StatusCode, err
```

File: go_lib/module/docs-builder/client.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
```
Comment (nabokihms) [inline go_lib/module/docs-builder/client.go:114]: NIT
```suggestion
		return nil, res.StatusCode, err
```

File: modules/810-documentation/templates/deployment.yaml [hunk 6]
```diff
@@ -49,6 +51,60 @@ spec:
...
+            path: /healthz
+            port: 8081
+        resources:
+          requests:
+            {{- include "helm_lib_module_ephemeral_storage_logs_with_extra" 10 | nindent 12 }}
+  {{- if not ( .Values.global.enabledModules | has "vertical-pod-autoscaler-crd") }}
+            {{- include "builder_resources" . | nindent 12 }}
+  {{- end }}
+        securityContext:
+          allowPrivilegeEscalation: false
```
Comment (nabokihms) [inline modules/810-documentation/templates/deployment.yaml:100]:
```suggestion
        securityContext:
          allowPrivilegeEscalation: false
```
This is required to use one of the following helpers https://github.com/deckhouse/lib-helm/tree/main/charts/helm_lib#module-security-context

You can find examples in the other modules in the repo.

**Review by nabokihms — CHANGES_REQUESTED** [review]

**Comment by z9r5** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.25 7570703932 7570703932-1-con-1-25 /sys/deckhouse-oss/install:pr6449 centos@3.66.249.70

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've recognized your '/e2e/abort' command and started the workflow...

File: testing/matrix/linter/rules/roles/placement.go [hunk 7]
```diff
@@ -38,7 +38,12 @@ func isSystemNamespace(actual string) bool {
 }
 
 func isDeckhouseSystemNamespace(actual string) bool {
-	return actual == "d8-monitoring" || actual == "d8-system"
+	return actual == "d8-monitoring" ||
+		actual == "d8-system" ||
+		//TODO: remove next lines after RBAC objects naming fixes
+		actual == "d8-admission-policy-engine" ||
+		actual == "d8-operator-trivy" ||
+		actual == "d8-log-shipper"
```
[the remark this answers is not in the document]
Comment (libmonsoon-dev) [reply testing/matrix/linter/rules/roles/placement.go]: I've added more details in the comment.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/aws'...

File: modules/810-documentation/templates/deployment.yaml (hunk 6 above)
Comment (libmonsoon-dev) [reply modules/810-documentation/templates/deployment.yaml]: Fixed

File: deckhouse-controller/pkg/controller/module-controllers/downloader/module_downloader.go (hunk 3 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/module-controllers/downloader/module_downloader.go]: This method is no longer used, I have removed it

File: modules/810-documentation/templates/vertical-pod-autoscaler.yaml [hunk 8]
```diff
@@ -27,4 +32,10 @@ spec:
       maxAllowed:
         cpu: 200m
         memory: 100Mi
+    - containerName: "builder"
+      minAllowed:
+        {{- include "builder_resources" . | nindent 8 }}
+      maxAllowed:
+        cpu: "1"
+        memory: 500Mi
```
[the remark this answers is not in the document]
Comment (libmonsoon-dev) [reply modules/810-documentation/templates/vertical-pod-autoscaler.yaml]: Fixed

File: go_lib/module/docs-builder/client.go (hunk 4 above)
Comment (libmonsoon-dev) [reply go_lib/module/docs-builder/client.go]: Fixed

File: deckhouse-controller/pkg/controller/module-controllers/downloader/module_downloader.go [hunk 9]
```diff
@@ -348,6 +362,15 @@ func (md *ModuleDownloader) fetchModuleReleaseMetadata(img v1.Image) (moduleRele
 	return meta, err
 }
 
+func (md *ModuleDownloader) ListModules() ([]string, error) {
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/module-controllers/downloader/module_downloader.go:365]: What for do we need this? I don't see any usages

**Review by MariaMakeeva — APPROVED** [review]

**Review by yalosev — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Pull request merged**